Mathlib/Tactic/Module.lean

```diff
@@ -0,0 +1,641 @@
+
+@[simp] theorem smul_apply [Mul R] (r : R) (l : NF R M) : r • l = l.map fun (a, x) ↦ (r * a, x) :=
+  rfl
+
+theorem eval_smul [AddCommMonoid M] [Semiring R] [Module R M] {l : NF R M} {x : M} (h : x = l.eval)
+    (r : R) : (r • l).eval = r • x := by
+  unfold NF.eval at h ⊢
+  simp only [h, smul_sum, map_map, NF.smul_apply]
+  congr
+  ext p
+  simp [mul_smul]
+
+theorem smul_eq_eval {R₀ : Type*} [AddCommMonoid M] [Semiring R] [Module R M] [Semiring R₀]
+    [Module R₀ M] [Semiring S] [Module S M]  {l : NF R M} {l₀ : NF R₀ M} {s : S} {r : R}
+    {x : M} (hx : x = l₀.eval) (hl : l.eval = l₀.eval) (hs : r • x = s • x) :
+    s • x = (r • l).eval := by
+  rw [← hs, hx, ← hl, eval_smul]
+  rfl
+
+theorem eq_cons_cons [AddMonoid M] [SMul R M] {r₁ r₂ : R} (m : M) {l₁ l₂ : NF R M} (h1 : r₁ = r₂)
+    (h2 : l₁.eval = l₂.eval) :
+    ((r₁, m) ::ᵣ l₁).eval = ((r₂, m) ::ᵣ l₂).eval := by
+  simp only [NF.eval, NF.cons] at *
+  simp [h1, h2]
+
+theorem eq_cons_const [AddCommMonoid M] [Semiring R] [Module R M] {r : R} (m : M) {n : M}
+    {l : NF R M} (h1 : r = 0) (h2 : l.eval = n) :
+    ((r, m) ::ᵣ l).eval = n := by
+  simp only [NF.eval, NF.cons] at *
+  simp [h1, h2]
+
+theorem eq_const_cons [AddCommMonoid M] [Semiring R] [Module R M] {r : R} (m : M) {n : M}
+    {l : NF R M} (h1 : 0 = r) (h2 : n = l.eval) :
+    n = ((r, m) ::ᵣ l).eval := by
+  simp only [NF.eval, NF.cons] at *
+  simp [← h1, h2]
+
+theorem eq_of_eval_eq_eval {R₁ R₂ : Type*} [AddCommMonoid M] [Semiring R] [Module R M] [Semiring R₁]
+    [Module R₁ M] [Semiring R₂] [Module R₂ M]  {l₁ l₂ : NF R M} {l₁' : NF R₁ M} {l₂' : NF R₂ M}
+    {x₁ x₂ : M} (hx₁ : x₁ = l₁'.eval) (hx₂ : x₂ = l₂'.eval) (h₁ : l₁.eval = l₁'.eval)
+    (h₂ : l₂.eval = l₂'.eval) (h : l₁.eval = l₂.eval) :
+    x₁ = x₂ := by
+  rw [hx₁, hx₂, ← h₁, ← h₂, h]
+
+variable (R)
+
+/-- Operate on a `Module.NF S M` object `l`, i.e. a list of pairs in `S × M`, where `S` is some
+commutative semiring, by applying to each `S`-component the algebra-map from `S` into a specified
+`S`-algebra `R`. -/
+def algebraMap [CommSemiring S] [Semiring R] [Algebra S R] (l : NF S M) : NF R M :=
+  l.map (fun ⟨s, x⟩ ↦ (_root_.algebraMap S R s, x))
+
+theorem eval_algebraMap [CommSemiring S] [Semiring R] [Algebra S R] [AddMonoid M] [SMul S M]
+    [MulAction R M] [IsScalarTower S R M] (l : NF S M) :
+    (l.algebraMap R).eval = l.eval := by
+  simp only [NF.eval, algebraMap, map_map]
+  congr
+  ext
+  simp [IsScalarTower.algebraMap_smul]
+
+end NF
+
+variable {u v : Level}
+
+/-! ### Lists of expressions representing scalars and vectors, and operations on such lists -/
+
+/-- Basic meta-code object of the `match_scalars` and `module` tactics: a type synonym for a list of
+ordered triples comprising expressions representing terms of two types `R` and `M` (where typically
+`M` is an `R`-module), together with a natural number "index".
+
+The natural number represents the index of the `M` term in the `AtomM` monad: this is not enforced,
```

Comment:
Just noting that we found with `omega` that the `AtomM` monad can result in bad performance on large problems. Leo wrote `Lean.Meta.Canonicalizer.canon` to solve this problem.

Reply:
#16621 for future discoverability!

Not expecting you to change anything here. (Although perhaps add a TODO in case someone runs into this issue?)

Reply:
As far as I can tell, `CanonM` is not a drop-in replacement for `AtomM`, for questions I have now brought up on that PR.  However, one observation from reading the [`CanonM` file](https://github.com/leanprover/lean4/blob/master/src/Lean/Meta/Canonicalizer.lean) is that Leo writes,
> The definitionally equality tests are performed using the given transparency mode. We claim `TransparencyMode.instances` is a good setting for most applications.

I'd be happy to downgrade from `.default` to `.instances` in this PR, do you think that would be a good change?  Honestly I sometimes find `ring`'s unwrapping ability (at `.default`) to be a little surprising.

Reply:
After canonicalizing via `CanonM`, it suffices to just use a `HashMap Expr Nat` to keep track of previously seen atoms.

This is strictly weaker than using AtomM (equality after canonicalization is not the same as defeq), but it is much faster on large problems. For `omega` this was a clear win.

Reply:
@semorrison Thanks, I believe I understand the `CanonM` paradigm now!

Sticking to `AtomM` now for simplicity: above I started to suspect that I should be working in this tactic at `.instances` rather than `.default` -- do you have an opinion?

Reply:
@hrmacbeth how would you feel about adding a note in the code (not the docstring) like this: "if poor performance on large problems is witnessed, consider switching the implementation to `CanonM` rather than `AtomM`, per this discussion: https://github.com/leanprover-community/mathlib4/pull/16593/files#r1749623191"